File: aat-lib/src/main/java/ch/bailu/aat_lib/service/directory/IteratorAbstract.java
package ch.bailu.aat_lib.service.directory;

import javax.annotation.Nonnull;

import ch.bailu.aat_lib.app.AppContext;
import ch.bailu.aat_lib.dispatcher.AppBroadcaster;
import ch.bailu.aat_lib.dispatcher.BroadcastReceiver;
import ch.bailu.aat_lib.gpx.GpxInformation;
import ch.bailu.aat_lib.preferences.OnPreferencesChanged;
import ch.bailu.aat_lib.preferences.SolidDirectoryQuery;
import ch.bailu.aat_lib.preferences.StorageInterface;
import ch.bailu.aat_lib.util.sql.DbResultSet;
import ch.bailu.foc.Foc;

public abstract class IteratorAbstract extends Iterator implements OnPreferencesChanged {

    private OnCursorChangedListener onCursorChangedListener = NULL_LISTENER;
    private DbResultSet resultSet = null;
    private final SolidDirectoryQuery sdirectory;
    private String selection="";
    private final BroadcastReceiver onSyncChanged = objs -> query();

    private final AppContext appContext;


    public IteratorAbstract (AppContext appContext) {
        this.appContext = appContext;
        sdirectory = new SolidDirectoryQuery(appContext.getStorage(), appContext);
        sdirectory.register(this);
        appContext.getBroadcaster().register(onSyncChanged, AppBroadcaster.DB_SYNC_CHANGED);
        openAndQuery();
    }


    @Override
    public void setOnCursorChangedListener(OnCursorChangedListener l) {
        onCursorChangedListener = l;
    }


    @Override
    public void onPreferencesChanged(@Nonnull StorageInterface s, @Nonnull String key) {
        if (sdirectory.hasKey(key)) {
            openAndQuery();

        } else if (sdirectory.containsKey(key) && !selection.equals(sdirectory.createSelectionString())) {
            query();
        }
    }


    @Override
    public boolean moveToPrevious() {
        if (resultSet!=null) return resultSet.moveToPrevious();
        return false;
    }


    @Override
    public boolean moveToNext() {
        if (resultSet != null) return resultSet.moveToNext();
        return false;
    }


    @Override
    public boolean moveToPosition(int pos) {
        if (resultSet != null) return resultSet.moveToPosition(pos);
        return false;
    }


    @Override
    public int getCount() {
        if (resultSet != null) return resultSet.getCount();
        return 0;
    }


    @Override
    public int getPosition() {
        if (resultSet != null) return resultSet.getPosition();
        return -1;
    }


    @Override
    public abstract GpxInformation getInfo();

    public abstract void onCursorChanged(DbResultSet resultSet, Foc directory, String fid);


    private void openAndQuery() {
        String fileOnOldPosition = "";
        int oldPosition=0;

        appContext.getServices().getDirectoryService().openDir(sdirectory.getValueAsFile());
        selection = sdirectory.createSelectionString();
        resultSet = appContext.getServices().getDirectoryService().query(selection);

        if (resultSet != null) {
            resultSet.moveToPosition(oldPosition);

            onCursorChanged(resultSet, sdirectory.getValueAsFile(), fileOnOldPosition);
            onCursorChangedListener.onCursorChanged();
        }
    }

    @Override
    public void query() {
        String fileOnOldPosition = "";
        int oldPosition=0;

        if (resultSet != null) {
            oldPosition = resultSet.getPosition();
            fileOnOldPosition = getInfo().getFile().getPath();
            resultSet.close();
        }

        selection = sdirectory.createSelectionString();
        resultSet = appContext.getServices().getDirectoryService().query(selection);

        if (resultSet != null) {
            resultSet.moveToPosition(oldPosition);

            onCursorChanged(resultSet, sdirectory.getValueAsFile(), fileOnOldPosition);
            onCursorChangedListener.onCursorChanged();
        }
    }

    @Override
    public void close() {
        if (resultSet!= null) resultSet.close();
        sdirectory.unregister(this);
        appContext.getBroadcaster().unregister(onSyncChanged);
        onCursorChangedListener = NULL_LISTENER;
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/service/directory/IteratorAbstract.kt
package ch.bailu.aat_lib.service.directory

import ch.bailu.aat_lib.app.AppContext
import ch.bailu.aat_lib.broadcaster.AppBroadcaster
import ch.bailu.aat_lib.broadcaster.BroadcastReceiver
import ch.bailu.aat_lib.logger.AppLog
import ch.bailu.aat_lib.preferences.OnPreferencesChanged
import ch.bailu.aat_lib.preferences.StorageInterface
import ch.bailu.aat_lib.preferences.file_list.SolidDirectoryQuery
import ch.bailu.aat_lib.service.directory.database.GpxDbConfiguration
import ch.bailu.aat_lib.util.sql.DbResultSet
import ch.bailu.foc.Foc


abstract class IteratorAbstract(private val appContext: AppContext) : Iterator(),
    OnPreferencesChanged {
    private var onCursorChangedListener = { }
    private var resultSet: DbResultSet? = null
    private val sdirectory: SolidDirectoryQuery = SolidDirectoryQuery(appContext.storage, appContext)
    private var extraStatement = sdirectory.createExtraStatement()
    private var extraStatementParam = sdirectory.createExtraStatementParam()
    private val onSyncChanged = BroadcastReceiver { _: Array<out String> -> query() }

    init {
        sdirectory.register(this)
        appContext.broadcaster.register(AppBroadcaster.DB_SYNC_CHANGED, onSyncChanged)
        openAndQuery()
    }

    override fun setOnCursorChangedListener(listner: () -> Unit) {
        onCursorChangedListener = listner
    }

    override fun onPreferencesChanged(storage: StorageInterface, key: String) {
        if (sdirectory.hasKey(key)) {
            extraStatement = sdirectory.createExtraStatement()
            extraStatementParam = sdirectory.createExtraStatementParam()
            openAndQuery()
        } else if (sdirectory.containsKey(key)) {
            val newExtraStatement = sdirectory.createExtraStatement()
            val newExtraStatementParam = sdirectory.createExtraStatementParam()
            if (extraStatement != newExtraStatement || extraStatementParam != newExtraStatementParam) {
                extraStatement = newExtraStatement
                extraStatementParam = newExtraStatementParam
                query()
            }
        }
    }

    override fun moveToPrevious(): Boolean {
        return resultSet?.moveToPrevious() ?: false
    }

    override fun moveToNext(): Boolean {
        return resultSet?.moveToNext() ?: false
    }

    override fun moveToPosition(pos: Int): Boolean {
        return resultSet?.moveToPosition(pos) ?: false
    }

    override fun getID(): Long {
        resultSet?.apply {
            return getLong(GpxDbConfiguration.ATTR_ID)
        }
        return 0L
    }

    override fun getCount(): Int {
        return resultSet?.getCount() ?: 0
    }

    override fun getPosition(): Int {
        return resultSet?.getPosition() ?: -1
    }

    abstract fun onCursorChanged(resultSet: DbResultSet, directory: Foc, fileID: String)

    private fun openAndQuery() {
        val fileOnOldPosition = ""
        val oldPosition = 0
        appContext.services.getDirectoryService().openDir(sdirectory.getValueAsFile())
        updateResultFromSelection()
        moveToOldPosition(oldPosition, fileOnOldPosition)
    }

    override fun query() {
        var fileOnOldPosition = ""
        var oldPosition = 0
        val resultSet = resultSet
        if (resultSet is DbResultSet) {
            oldPosition = resultSet.getPosition()
            fileOnOldPosition = getInfo().getFile().path
            resultSet.close()
        }
        updateResultFromSelection()
        moveToOldPosition(oldPosition, fileOnOldPosition)
    }

    private fun updateResultFromSelection() {
        try {
            if (extraStatementParam.isEmpty()) {
                resultSet = appContext.services.getDirectoryService()
                    .select(extraStatement)
            } else {
                resultSet = appContext.services.getDirectoryService()
                    .select(extraStatement, extraStatementParam)
            }
        } catch (e: Exception) {
            AppLog.e(this, e)
        }
    }

    private fun moveToOldPosition(oldPosition: Int, fileOnOldPosition: String) {
        val resultSet = resultSet
        if (resultSet is DbResultSet) {
            resultSet.moveToPosition(oldPosition)
            onCursorChanged(resultSet, sdirectory.getValueAsFile(), fileOnOldPosition)
            onCursorChangedListener()
        }
    }

    override fun close() {
        resultSet?.close()
        sdirectory.unregister(this)
        appContext.broadcaster.unregister(onSyncChanged)
        onCursorChangedListener = { }
    }
}
